static/extensions/Ikelene/email2FA.js

```diff
@@ -0,0 +1,317 @@
+                    //    arguments: {
+                    //        EMAIL: { type: Scratch.ArgumentType.STRING, defaultValue: 'your@email.com' },
+                    //        PASS: { type: Scratch.ArgumentType.STRING, defaultValue: 'password' },
+                    //        SMTP: { type: Scratch.ArgumentType.STRING, defaultValue: 'smtp.gmail.com' },
+                    //        PORT: { type: Scratch.ArgumentType.STRING, defaultValue: '465' }
+                    //    }
+                    //},
+                    //{
+                    //    opcode: 'clearCustomEmailLogin',
+                    //    blockType: Scratch.BlockType.COMMAND,
+                    //    text: 'Clear custom email login'
+                    //}
+                ],
+                menus: {
+                    presetMenu: {
+                        acceptReporters: true,
+                        items: [
+                            { text: 'Classic (default)', value: '5' },
+ { text: 'Yellow Card', value: '1' },
+ { text: 'Blue Card', value: '2' },
+ { text: 'Minimalist Dashed', value: '3' },
+ { text: 'Welcome Style', value: '4' }
+                        ]
+                    }
+                }
+            };
+        }
+
+        constructor() {
+            const runtime = Scratch.runtime;
+            if (runtime && runtime.vm) {
+                const state = runtime.vm.extensionState['IkeEmailTwofa'] ||= {};
+                cachedProjectID = state['projectID'] || null;
+            } else {
+                console.warn('Scratch.runtime.vm not available yet.');
+            }
+        }
+
+        setEmailPreset(args) {
+            emailPreset = parseInt(args.PRESET);
+        }
+
+        setCustomHTML(args) {
+            customHTML = args.HTML;
+        }
+
+        //setCustomEmailLogin(args) {
+        //    customEmailSettings = {
+        //        email: args.EMAIL,
+        //        pass: args.PASS,
+        //        smtp: args.SMTP,
+        //        port: args.PORT
+        //    };
+        //}
+
+        //clearCustomEmailLogin() {
+        //    customEmailSettings = null;
+        //}
+
+        async registerProjectID(args) {
+            if (cachedProjectID) return cachedProjectID;
+
+            try {
+                const res = await fetch('https://ikelene.ca/api/2fa/register_project.php', {
+                    method: 'POST',
+                    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+                    body: `project=${encodeURIComponent(args.PROJECTNAME)}`
+                });
+
+                const id = await res.text();
+                const trimmed = id.trim();
+
+                const isError = ['invalid name', 'invalid chars', 'rate limit', 'rate limited'].includes(trimmed);
+                if (trimmed && !isError) {
+                    cachedProjectID = trimmed;
+
+                    const runtime = Scratch.runtime;
```

Comment:
Store your runtime directly in your extension by adding it on construction rather than getting it from the sandbox every single time. Also, I think I've already said that it's passed into the constructor as the first argument.

Reply:
```js
constructor(runtime) {
  this.runtime = runtime;
}
```

Reply:
Actually, nevermind.